### PR TITLE
RIA-2794: set up docker for HOIS API

### PR DESCRIPTION
**Pull request by divisathyan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RIA-2794

### Change description ###
Initial project set up. Created a dockerized Spring boot app with REST API, Swagger

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: Jenkinsfile_nightly [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#!groovy
```
Comment (raghera) [inline Jenkinsfile_nightly]: Do we need a Jenkins_CNP file?

File: Jenkinsfile_nightly [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    secret('idam-secret', 'IA_IDAM_SECRET'),
+    secret('idam-redirect-uri', 'IA_IDAM_REDIRECT_URI'),
+    secret('s2s-secret', 'IA_S2S_SECRET'),
+    secret('s2s-microservice', 'IA_S2S_MICROSERVICE'),
+    secret('ccd-url', 'CCD_URL'),
+    secret('ccd-gw-url', 'CCD_GW_URL'),
+    secret('dm-url', 'DM_URL'),
+    secret('idam-url', 'IDAM_URL'),
+    secret('s2s-url', 'S2S_URL'),
+    secret('prof-ref-data-url', 'PROF_REF_DATA_URL')
```
Comment (raghera) [inline Jenkinsfile_nightly]: We should only add what we need here.  IDAM and S2S is ok.  The rest we can add as we need them.

File: LICENSE [hunk 3]
```diff
@@ -1,21 +1,9 @@
...
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (raghera) [inline LICENSE]: seems to be only formatting changes here?

File: README.md [hunk 4]
```diff
@@ -1 +1,4 @@
-# ia-home-office-integration-api
+# Home Office Integration Service API
```
Comment (raghera) [inline README.md]: Let's just call it the Home Office Integration API

File: bin/run-in-docker.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  Available parameters:
+
+  "
+}
+
+# script execution flags
+GRADLE_CLEAN=false
+GRADLE_INSTALL=false
+
+# TODO custom environment variables application requires.
```
Comment (raghera) [inline bin/run-in-docker.sh]: Ok for now as it's just first commit, but let's make sure we remove any TODOs

File: build.gradle [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+  compile group: 'com.github.ben-manes.caffeine', name: 'caffeine', version: '2.5.6'
+  compile group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.10.0.pr3'
+  compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.10.0.pr3'
+  compile group: 'com.fasterxml.jackson.datatype', name: 'jackson-datatype-jsr310', version: '2.10.0.pr2'
+  compile group: 'io.github.openfeign.form', name: 'feign-form', version: '3.4.0'
+  compile group: 'io.github.openfeign.form', name: 'feign-form-spring', version: '3.4.0'
+  compile group: 'io.springfox', name: 'springfox-swagger2', version: versions.springfoxSwagger
+  compile group: 'io.springfox', name: 'springfox-swagger-ui', version: versions.springfoxSwagger
+  //compile group: 'org.flywaydb', name: 'flyway-core', version: '5.2.4'
```
Comment (raghera) [inline build.gradle]: no DB so don't need this.

File: build.gradle [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+  //compile group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '3.0.0'
+  //compile(group: 'uk.gov.hmcts.reform.auth', name: 'auth-checker-lib', version: '2.1.4')
+  compile group: 'com.google.guava', name: 'guava', version: '27.0-jre'
+  compile 'org.projectlombok:lombok:1.18.2'
+
+  compile group: 'com.github.ben-manes.caffeine', name: 'caffeine', version: '2.5.6'
+  compile group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.10.0.pr3'
+  compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.10.0.pr3'
+  compile group: 'com.fasterxml.jackson.datatype', name: 'jackson-datatype-jsr310', version: '2.10.0.pr2'
```
Comment (raghera) [inline build.gradle]: can bump jackson to the latest

File: charts/ia-homeoffice-integration-api/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+name: ia-homeoffice-integration-service
+home: https://github.com/divisathyan/ia-ho-integration-services
```
Comment (raghera) [inline charts/ia-homeoffice-integration-api/Chart.yaml]: please make sure this is pointing to the hmcts repo

File: charts/ia-homeoffice-integration-api/values.yaml [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        - idam-redirect-uri
+        - s2s-secret
+        - s2s-microservice
+        - ccd-url
+        - ccd-gw-url
+        - dm-url
+        - idam-url
+        - s2s-url
+        - prof-ref-data-urls
+        - AppInsightsInstrumentationKey
```
Comment (raghera) [inline charts/ia-homeoffice-integration-api/values.yaml]: Let's get rid of everything we know don't need and add as required.  idam and s2s we will need, the rest probably not.  We will need `AppInsightsIntrumentationKey`

File: config/owasp/suppressions.xml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  <suppress until="2030-01-01">
+    <notes><![CDATA[
+     Suppressing as it's a false positive (see: https://pivotal.io/security/cve-2018-1258)
+   ]]></notes>
+    <gav regex="true">^org\.springframework\.security:spring-security-crypto:5.2.[0-2].RELEASE</gav>
+    <cpe>cpe:/a:pivotal_software:spring_security</cpe>
+    <cve>CVE-2018-1258</cve>
+  </suppress>
+
+  <suppress until="2020-01-31">
```
Comment (raghera) [inline config/owasp/suppressions.xml]: Is this suppressions still required?

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+          DM_URL: http://dm-store:8080
+          IA_IDAM_CLIENT_ID:
+          IA_IDAM_SECRET:
+          IA_S2S_MICROSERVICE:
+          IA_S2S_SECRET:
+          IA_SYSTEM_USERNAME:
+          IA_SYSTEM_PASSWORD:
+          IA_EM_STITCHING_ENABLED:
+          IA_DOCMOSIS_ENABLED:
+          IA_SUBMIT_HEARING_REQUIREMENTS_ENABLED:
```
Comment (raghera) [inline docker-compose.yml]: please remove things we don't know we need.

File: infrastructure/aat.tfvars [hunk 12]
```diff
@@ -0,0 +1,2 @@
+capacity = "1"
```
Comment (raghera) [inline infrastructure/aat.tfvars]: let's keep for now ... but don't think we need this any more

File: infrastructure/demo.tfvars [hunk 13]
```diff
@@ -0,0 +1,3 @@
+capacity = "1"
+instance_size="I1"
+enable_ase = true
```
Comment (raghera) [inline infrastructure/demo.tfvars]: remove the `enable_ase` flag.  We def don't need that.

File: src/integrationTest/java/uk/gov/hmcts/reform/hois/controllers/HoisControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import uk.gov.hmcts.reform.hois.domain.entities.HomeOfficeRequest;
+
+@WebMvcTest
```
Comment (raghera) [inline src/integrationTest/java/uk/gov/hmcts/reform/hois/controllers/HoisControllerTest.java]: Probably better to use something like 
```@RunWith(SpringRunner.class)
@SpringBootTest(classes = Application.class, webEnvironment = MOCK)```

See ia-case-api for reference

File: Jenkinsfile_nightly (hunk 2 above)
Comment (divisathyan) [reply Jenkinsfile_nightly]: Removed other variables, kept only IDAM and S2S

File: LICENSE (hunk 3 above)
Comment (divisathyan) [reply LICENSE]: The license is from the new spring-boot template.

File: README.md (hunk 4 above)
Comment (divisathyan) [reply README.md]: Corrected

File: bin/run-in-docker.sh (hunk 5 above)
Comment (divisathyan) [reply bin/run-in-docker.sh]: It was already present in the template. I can remove if required

File: build.gradle (hunk 6 above)
Comment (divisathyan) [reply build.gradle]: It was commented and now removed

File: build.gradle (hunk 7 above)
Comment (divisathyan) [reply build.gradle]: spring boot compatible version is 2.10.2

File: charts/ia-homeoffice-integration-api/Chart.yaml (hunk 8 above)
Comment (divisathyan) [reply charts/ia-homeoffice-integration-api/Chart.yaml]: corrected

File: charts/ia-homeoffice-integration-api/values.yaml (hunk 9 above)
Comment (divisathyan) [reply charts/ia-homeoffice-integration-api/values.yaml]: removed other variables

File: config/owasp/suppressions.xml (hunk 10 above)
Comment (divisathyan) [reply config/owasp/suppressions.xml]: I checked the NVD and both of the CVE are in Assigned status. so, kept it in the file

File: docker-compose.yml (hunk 11 above)
Comment (divisathyan) [reply docker-compose.yml]: Removed other variables

File: infrastructure/aat.tfvars (hunk 12 above)
Comment (divisathyan) [reply infrastructure/aat.tfvars]: I have retained the same variables as ia-case-api

File: infrastructure/demo.tfvars (hunk 13 above)
Comment (divisathyan) [reply infrastructure/demo.tfvars]: removed

File: src/integrationTest/java/uk/gov/hmcts/reform/hois/controllers/HoisControllerTest.java (hunk 14 above)
Comment (divisathyan) [reply src/integrationTest/java/uk/gov/hmcts/reform/hois/controllers/HoisControllerTest.java]: Do we need the Spring context to be set up for web test?

Comment (divisathyan) [reply src/integrationTest/java/uk/gov/hmcts/reform/hois/controllers/HoisControllerTest.java]: Changed to SpringBootTest

File: Jenkinsfile_nightly (hunk 1 above)
Comment (divisathyan) [reply Jenkinsfile_nightly]: copied from template

Comment (raghera) [reply Jenkinsfile_nightly]: Please add the jenkins_CNP file.

File: LICENSE (hunk 3 above)
Comment (raghera) [reply LICENSE]: Can we keep the one already in `master` since the year is up to date.

File: bin/run-in-docker.sh (hunk 5 above)
Comment (raghera) [reply bin/run-in-docker.sh]: Ok to leave for now as I said.

File: Jenkinsfile_nightly (hunk 1 above)
Comment (divisathyan) [reply Jenkinsfile_nightly]: Added

File: LICENSE (hunk 3 above)
Comment (divisathyan) [reply LICENSE]: Sure. Updated the file as in master

**Comment by raghera** [comment]
Build fails due to a bug in the script checking whether the `aadIdentityName` is in master.  The PR is correct so merging.

**Pull request merged**